tests/src/test/spark330/scala/org/apache/spark/sql/rapids/utils/RapidsTestsTrait.scala

```diff
@@ -0,0 +1,334 @@
+    }
+
+    if (
+      !(checkResult(result.head.get(0), expected, expression.dataType, expression.nullable)
+        || checkResult(
+          CatalystTypeConverters.createToCatalystConverter(expression.dataType)(
+            result.head.get(0)
+          ), // decimal precision is wrong from value
+          CatalystTypeConverters.convertToCatalyst(expected),
+          expression.dataType,
+          expression.nullable
+        ))
+    ) {
+      val input = if (inputRow == EmptyRow) "" else s", input: $inputRow"
+      fail(
+        s"Incorrect evaluation: $expression, " +
+          s"actual: ${result.head.get(0)}, " +
+          s"expected: $expected$input")
+    }
+  }
+
+  def shouldNotFallback(): Unit = {
+    TestStats.offloadRapids = false
+  }
+
+  def canConvertToDataFrame(inputRow: InternalRow): Boolean = {
+    if (inputRow == EmptyRow || inputRow == InternalRow.empty) {
+      return true
+    }
+    if (!inputRow.isInstanceOf[GenericInternalRow]) {
+      return false
+    }
+    val values = inputRow.asInstanceOf[GenericInternalRow].values
+    for (value <- values) {
+      value match {
+        case _: MapData => return false
+        case _: ArrayData => return false
+        case _: InternalRow => return false
+        case _ =>
+      }
+    }
+    true
+  }
+
+  def convertInternalRowToDataFrame(inputRow: InternalRow): DataFrame = {
+    val structFileSeq = new ArrayBuffer[StructField]()
+    val values = inputRow match {
+      case genericInternalRow: GenericInternalRow =>
+        genericInternalRow.values
+      case _ => throw new UnsupportedOperationException("Unsupported InternalRow.")
+    }
+    values.map {
+      case boolean: java.lang.Boolean =>
+        structFileSeq.append(StructField("bool", BooleanType, boolean == null))
+      case short: java.lang.Short =>
+        structFileSeq.append(StructField("i16", ShortType, short == null))
+      case byte: java.lang.Byte =>
+        structFileSeq.append(StructField("i8", ByteType, byte == null))
+      case integer: java.lang.Integer =>
+        structFileSeq.append(StructField("i32", IntegerType, integer == null))
+      case long: java.lang.Long =>
+        structFileSeq.append(StructField("i64", LongType, long == null))
+      case float: java.lang.Float =>
+        structFileSeq.append(StructField("fp32", FloatType, float == null))
+      case double: java.lang.Double =>
+        structFileSeq.append(StructField("fp64", DoubleType, double == null))
+      case utf8String: UTF8String =>
+        structFileSeq.append(StructField("str", StringType, utf8String == null))
+      case byteArr: Array[Byte] =>
+        structFileSeq.append(StructField("vbin", BinaryType, byteArr == null))
+      case decimal: Decimal =>
+        structFileSeq.append(
+          StructField("dec", DecimalType(decimal.precision, decimal.scale), decimal == null))
+      case _ =>
+        // for null
+        structFileSeq.append(StructField("n", IntegerType, nullable = true))
```

Comment:
refined the code to throw exception on other types.

As to the IntegerType question, I changed IntegerType to NullType and found org.apache.spark.sql.rapids.suites.RapidsStringExpressionsSuite will fail at compiling codegen codes. I think this is a workaround from the Gluten side and since it makes no big difference we can keep the workaournd?